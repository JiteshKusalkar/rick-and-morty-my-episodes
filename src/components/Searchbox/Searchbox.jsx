import React from 'react';
import AsyncSelect from 'react-select/async';
import { instance, getParams } from '../../api';

export const loadOptions = (inputValue, callback) => {
  const params = getParams({ name: inputValue });
  instance.get('character', { params }).then(response => {
    const options = response.data.results.map(character => ({
      ...character,
      value: character.id,
      label: character.name
    }));

    if (callback) {
      callback(options);
    }
  });
};

function Searchbox({ defaultOptions, onChange }) {
  return (
    <AsyncSelect
      className="react_select"
      classNamePrefix="react_select"
      value=""
      placeholder="Select character or type a name"
      defaultOptions={defaultOptions}
      onChange={onChange}
      isSearchable={true}
      loadOptions={loadOptions}
    />
  );
}

export default Searchbox;
