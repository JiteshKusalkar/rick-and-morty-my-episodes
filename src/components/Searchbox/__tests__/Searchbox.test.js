import MockAdapter from 'axios-mock-adapter';
import { shallow } from 'enzyme';
import React from 'react';

import Searchbox, { loadOptions } from '../Searchbox';
import { instance } from '../../../api';

// tslint:disable-next-line:no-require-imports
const unroll = require('unroll');
unroll.use(it);

const validProps = {
  loadOptions: jest.fn(),
  defaultOptions: [
    {
      label: 'Rick',
      value: 1
    }
  ]
};

const shallowInstance = props => shallow(<Searchbox {...props} />);

// Snapshot testing
describe('Searchbox', () => {
  unroll(
    'should match the snapshot for #propType',
    (done, args) => {
      const wrapper = shallowInstance(args.value);
      expect(wrapper).toMatchSnapshot();
      done();
    },
    [
      ['propType', 'value'],
      ['valid props', validProps]
    ]
  );
});

// Unit testing
describe('Searchbox', () => {
  it('should load options', () => {
    jest.clearAllMocks();
    const mock = new MockAdapter(instance);
    const url = 'character';
    const callback = jest.fn();
    const mockData = {
      results: [
        {
          id: 1,
          name: 'Rick'
        }
      ]
    };

    mock.onGet(url, { params: { name: 'rick' } }).reply(200, mockData);
    loadOptions('rick', callback);

    expect(callback).toBeDefined();
  });
});
