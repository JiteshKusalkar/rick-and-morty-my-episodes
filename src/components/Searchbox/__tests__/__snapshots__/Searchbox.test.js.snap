// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Searchbox should match the snapshot for "valid props" 1`] = `
<Async
  cacheOptions={false}
  className="react_select"
  classNamePrefix="react_select"
  defaultOptions={
    Array [
      Object {
        "label": "Rick",
        "value": 1,
      },
    ]
  }
  filterOption={null}
  isSearchable={true}
  loadOptions={[Function]}
  placeholder="Select character or type a name"
  value=""
/>
`;
